### PR TITLE
Runner sets residual args instead of append

### LINKED ISSUE #14541: Problem running `@main` methods with command line arguments
## Compiler version
3.1.1

## Minimized code

```Scala
// In file echoargs.scala
@main def echo(args: String*) =
  println(args.mkString(" "))
```

## Output
First compile
```
$ scalac echoargs.scala
```
Then run the main method:
```
$ scala echo hello world
echo hello world hello world
```

## Expectation
```
$ scala echo hello world
hello world
```
Edit: this was the output using the `scala` command of 3.0.2

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes #14541 

Quick tweak requires proper test.

Backported https://github.com/lampepfl/dotty/pull/14552

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Here is an example of how to call the `def main` that is generated by the `@main` in a test.

https://github.com/lampepfl/dotty/pull/13727/files#diff-4c9963dd435c987466fdd17e4c3c0bfa9d5bf45e42f1c505870902e5981438c4R8-R30

**Comment by smarter** [comment]
Would you mind opening a backport of this fix to the release-3.1.2 branch too?

**Comment by som-snytt** [comment]
The test doesn't help, and `MainGenericRunner` isn't amenable to testing. I know there are script-level tests, so I'll track that down. In Scala 2, it figures out a class path programmatically rather than scriptedly.

Meanwhile, I'll learn how to backport to the release branch.

**Comment by smarter** [comment]
You can add shell script unit tests to https://github.com/lampepfl/dotty/blob/main/project/scripts/cmdTests

**Comment by smarter** [comment]
Oh but cmdTests uses sbt scalac, I don't know what's the best place to test the output of the actual scalac shell script /cc @BarkingBad

**Comment by BarkingBad** [comment]
If I got you correctly, you are looking for these tests https://github.com/lampepfl/dotty/tree/main/compiler/test/dotty/tools/scripting
I think @philwalk is the best person to consult with since he applied the most recent changes to these tests

**Comment by som-snytt** [comment]
Thanks @BarkingBad and sorry everyone that I may sound like a complainer about dotty testing -- my background is that I began by complaining about partest, I mean I really hated it, but it evolved and I evolved and we reconciled -- but currently I feel like I need to spend a weekend to understand the test regimes just to write a test. My PR on `-Wunused;imports` got stuck on the test failing in CI for an undiagnosed reason. So, just to repeat, test frameworks are thankless work, and I appreciate the challenges and that there is an underlying reason I have trouble writing or running a test. So, thanks!

I think there is a process meme about "test code is also just code" in the sense that it bears the same costs as so-called regular code.

**Comment by nicolasstucki** [comment]
Adding the following test in [project/scripts/cmdTests](https://github.com/lampepfl/dotty/blob/main/project/scripts/cmdTests) does the appropriate test. Not sure if that is the place to use `bin/scala`
```bash
echo "testing bin/scala call to main"
clear_out "$OUT"
bin/scalac tests/run-with-compiler/i14541.scala
bin/scala echo hello world > "$tmp"
# cat "$tmp" # for debugging
grep -e "^hello world$" "$tmp"
```

File: tests/run-with-compiler/i14541.scala [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+object Test:
```
Comment (nicolasstucki) [inline tests/run-with-compiler/i14541.scala:2]: Found a way to properly test `MainGenericRunner.main`
```scala
@main def Test: Unit =
  val classpath = dotty.tools.dotc.util.ClasspathFromClassloader(getClass.getClassLoader)
  dotty.tools.MainGenericRunner.main(Array("-classpath", classpath, "echo", "hello", "world"))
```

Comment (nicolasstucki) [reply tests/run-with-compiler/i14541.scala:2]: @som-snytt could you double-check that works as expected with your change and fails without it?

Comment (nicolasstucki) [reply tests/run-with-compiler/i14541.scala:2]: @som-snytt did that work?

Comment (som-snytt) [reply tests/run-with-compiler/i14541.scala:2]: I'll try it out this morning.

**Comment by som-snytt** [comment]
Before
```
Output from 'tests/run-with-compiler/i14541.scala' did not match check file. Actual output:
echo hello world hello world
```
After
```
[success] Total time: 12 s, completed Feb 28, 2022, 3:18:03 AM
```

**Comment by som-snytt** [comment]
12 years ago
```
    /** I have distilled everyone's classpath hopes and dreams into the
     *  question of how to resolve this boolean.  Right at this moment if I
     *  do anything but default to true right here, partest chokes.  I'm
     *  steadily reworking all the places partest gets its hands on the
     *  classpath so eventually I'll be able to remedy that, at which point
     *  my current plan is to have this default to false unless some
     *  property or command line option is supplied.  This is negotiable,
     *  but at this point I can say with confidence that less magic and
     *  less autodetection is more better.
     */
    def useJavaClassPath    = true
```
The scala 2 script uses env var to set the default for the `-usejavacp` option
```
# scala/bug#8358, scala/bug#8368 -- the default should really be false,
# but I don't want to flip the default during 2.11's RC cycle
OVERRIDE_USEJAVACP="-Dscala.usejavacp=true"
```
At the command line, I'd expect to get a useful class path for free. But I haven't looked to see what assumptions or expectations are built into `scala-cli`, for contrast.

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 2]
```diff
@@ -192,18 +193,20 @@ object MainGenericRunner {
 
       case ExecuteMode.PossibleRun =>
         val newClasspath = (settings.classPath :+ ".").flatMap(_.split(classpathSeparator).filter(_.nonEmpty)).map(File(_).toURI.toURL)
-        import dotty.tools.runner.RichClassLoader._
-        val newClassLoader = ScalaClassLoader.fromURLsParallelCapable(newClasspath)
+        import dotty.tools.runner.RichClassLoader.{getClass as _, *}
+        val newClassLoader = ScalaClassLoader.fromURLsParallelCapable(newClasspath, getClass.getClassLoader)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/MainGenericRunner.scala:197]: This change does not seem to be needed to make `tests/run-with-compiler/i14541.scala` pass. Why was this added? How could I test this?

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 3]
```diff
@@ -166,9 +166,10 @@ object MainGenericRunner {
           .withTargetScript(arg)
           .withScriptArgs(tail*)
       else
-        val newSettings = if arg.startsWith("-") then settings else settings.withPossibleEntryPaths(arg).withModeShouldBePossibleRun
-        process(tail, newSettings.withResidualArgs(arg))
-
+        val modalSettings = if arg.startsWith("-") then settings else settings.withPossibleEntryPaths(arg).withModeShouldBePossibleRun
+        val newSettings = modalSettings.withResidualArgs(arg)
+        process(tail, newSettings)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/MainGenericRunner.scala:171]: This change does not seem to be needed to make `tests/run-with-compiler/i14541.scala` pass. Why was this added? How could I test this?

File: tests/run-with-compiler/i14541.scala (hunk 1 above)
Comment (nicolasstucki) [reply tests/run-with-compiler/i14541.scala:3]: Nice.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 3 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/MainGenericRunner.scala:171]: Just refactors the expression so I could println settings. I can revert.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 2 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/MainGenericRunner.scala:197]: This is the 2nd commit about whether to do `-usejavacp` behavior from Scala 2, i.e., use the class loader you were given. I'm doubtful both ways.

File: tests/run-with-compiler/i14541.scala [hunk 4]
```diff
@@ -0,0 +1,14 @@
+
+// test argument processing and "execution mode" in runner
+object Test:
+  import dotty.tools.runner.RichClassLoader.*
+  val classpath = dotty.tools.dotc.util.ClasspathFromClassloader(getClass.getClassLoader)
+  def main(args: Array[String]): Unit =
+    getClass.getClassLoader.run("echo", List("hello", "raw", "world"))
+    getClass.getClassLoader.run("dotty.tools.MainGenericRunner", List("echo", "hello", "rawr", "world"))
```
Comment (som-snytt) [inline tests/run-with-compiler/i14541.scala:8]: This call works after using "parent class loader" in Runner. Folks with an opinion about how the command line runner should behave might have an opinion.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 3 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/MainGenericRunner.scala:171]: We need a test tool that proves a local refactor cannot change behavior.

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 5]
```diff
@@ -199,11 +200,13 @@ object MainGenericRunner {
             Option.when(Jar.isJarOrZip(dotty.tools.io.Path(entryPath)))(Jar(entryPath).mainClass).flatten
           }.isDefined
         }
-        targetToRun match
+        val newSettings =
+          targetToRun match
```
Comment (som-snytt) [inline compiler/src/dotty/tools/MainGenericRunner.scala:204]: This indentation was also to facilitate `println` and can be reverted. It's kind of cool that the indentation works as advertised.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/MainGenericRunner.scala:196]: If that is the case, I would remove this change and the second test. The third test already covers the fix. We could have a look at this cp change in a separate PR as we don't know or have time to test unintended consequences of this change. I would prefer to only fix the regression in this PR as this fill be back ported. We also want the backport quickly to be able to release 3.1.2 as soon as possible.

Comment (smarter) [reply compiler/src/dotty/tools/MainGenericRunner.scala:196]: @som-snytt we also have a -usejavacp flag checked in some of our she'll scripts I believe, shouldn't we check it here too?

**Comment by som-snytt** [comment]
Dropped the speculative commit and squashed.

Thanks, for future contributions I'll be more disciplined about minimized and easy-to-review deltas.

**Review by smarter — APPROVED** [review]

**Review by nicolasstucki — APPROVED** [review]
